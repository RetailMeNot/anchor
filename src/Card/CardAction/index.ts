export { CardAction } from './CardAction.component';
