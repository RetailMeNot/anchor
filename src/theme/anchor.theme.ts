// ANCHOR THEME
import { radii } from './sizes.theme';
import { MODAL_THEME_KEY, MODAL_THEME } from '../Modal';

// Copied from Adapter.StandardBreakpoints mins
// todo: dedup- they should use a single source
const breakpoints = {
    sm: '576px',
    md: '768px',
    lg: '992px',
    xl: '1200px',
};

interface AnchorTheme {
    colors?: any;
    breakpoints?: any;
    radii?: any;
    [key: string]: any;
}

export const anchorTheme: AnchorTheme = {
    breakpoints,
    radii,
    [MODAL_THEME_KEY]: MODAL_THEME,
};
