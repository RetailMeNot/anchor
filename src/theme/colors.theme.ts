// VENDOR
import { transparentize } from 'polished';

export type Color =
    | 'grayscale'
    | 'purple'
    | 'blue'
    | 'green'
    | 'yellow'
    | 'red';

const colors = {
    grayscale: {
        100: '#000000',
        200: '#323232',
        300: '#595959',
        400: '#808080',
        500: '#A6A6A6',
        600: '#D3D3D3',
        700: '#E7E7E7',
        800: '#F1F1F1',
        900: '#FAFAFA',
    },
    purple: {
        base: '#602D6C',
        light: '#91609C',
        dark: '#4B2354',
    },
    blue: {
        base: '#0998D6',
        light: '#66CCFF',
        dark: '#0074A6',
    },
    green: {
        base: '#14A382',
        light: '#00CDAD',
        dark: '#268068',
    },
    yellow: {
        base: '#F5A623',
        light: '#FFC761',
        dark: '#CC861D',
    },
    red: {
        base: '#EA4545',
        light: '#F16667',
        dark: '#CC3B3B',
    },
};

type ColorHue = { light: string; base: string; dark: string };

export interface ColorsThemeSchema {
    primary: ColorHue;
    secondary: ColorHue;
    accent: ColorHue;

    borders: {
        base: string;
        light: string;
        dark: string;
        error: string;
        success: string;
        info: string;
        warning: string;
    };

    text: {
        base: string;
        light: string;
        label: string;
        link: {
            base: string;
            hover: string;
            focus: string;
            visited: string;
        };
        disabled: string;
        placeholder: string;
    };

    background: {
        base: string;
        light: string;
        content: string;
        error: string;
        success: string;
        info: string;
        warning: string;
    };

    error: string;
    success: string;
    info: string;
    warning: string;
}

export const ColorsTheme: ColorsThemeSchema = {
    primary: colors.purple,
    secondary: colors.green,
    accent: colors.blue,
    borders: {
        // PRAGMATIC
        light: colors.grayscale[700],
        base: colors.grayscale[600],
        dark: colors.grayscale[400],
        // SEMANTIC
        error: transparentize(0.4, colors.red.dark),
        success: transparentize(0.4, colors.green.base),
        info: transparentize(0.4, colors.blue.light),
        warning: transparentize(0.4, colors.yellow.light),
    },
    text: {
        // PRAGMATIC
        base: colors.grayscale[200],
        light: colors.grayscale[300],
        // SEMANTIC
        link: {
            base: colors.blue.dark,
            hover: colors.blue.dark,
            focus: colors.blue.dark,
            visited: colors.blue.dark,
        },
        disabled: colors.grayscale[600],
        placeholder: colors.grayscale[400],
        label: colors.grayscale[400],
    },
    background: {
        // PRAGMATIC
        base: colors.grayscale[800],
        light: colors.grayscale[900],
        content: 'white',
        // SEMANTIC
        error: transparentize(0.75, colors.red.dark),
        success: transparentize(0.75, colors.green.base),
        info: transparentize(0.75, colors.blue.light),
        warning: transparentize(0.75, colors.yellow.light),
    },
    error: colors.red.dark,
    success: colors.green.base,
    info: colors.blue.light,
    warning: colors.yellow.light,
};
