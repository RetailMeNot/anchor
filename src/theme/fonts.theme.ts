// VARIABLES
import variables from './variables.theme';

const { baseFontSize, fontFamily } = variables;

export default {
  baseFontSize,
  fontFamily,
  paragraph: {
    base: {
      fontSize: `${baseFontSize}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    informational: {
      fontSize: `${baseFontSize *  0.875}px`,
      lineHeight: `${baseFontSize * 1.125}px`,
      fontWeight: 'normal',
    },
  },
  navigation: {
    base: {
      fontSize: `${baseFontSize * .75}px`,
      lineHeight: `${baseFontSize}px`,
      fontWeight: '500',
    },
    menu: {
      fontSize: `${baseFontSize * .875}px`,
      lineHeight: `${baseFontSize}px`,
      small: {
        base: {
          fontSize: `${baseFontSize * 0.75}px`,
          fontWeight: '500',
        },
        active: {
          fontWeight: '700',
        },
      },
      large: {
        base: {
          fontWeight: '600',
        },
        active: {
          fontWeight: '700',
        },
      },
    },
  },
  heading: {
    h4: {
      fontSize: `${baseFontSize * 1.5}px`,
      lineHeight: `${baseFontSize * 2}px`,
      fontWeight: 'normal',
    },
    h3: {
      fontSize: `${baseFontSize * 1.75}px`,
      lineHeight: `${baseFontSize * 2}px`,
      fontWeight: 'normal',
    },
    h2: {
      fontSize: `${baseFontSize * 2}px`,
      lineHeight: `${baseFontSize * 2.5}px`,
      fontWeight: 'normal',
    },
    h1: {
      fontSize: `${baseFontSize * 2.25}px`,
      lineHeight: `${baseFontSize * 2.5}px`,
      fontWeight: 'normal',
    },
  },
  typography: {
    caption: {
      fontSize: `${baseFontSize * 0.75}px`,
      lineHeight: `${baseFontSize}px`,
      fontWeight: '500',
    },
    ui: {
      fontSize: `${baseFontSize * 0.875}px`,
      lineHeight: `${baseFontSize * 1.125}px`,
      fontWeight: 'normal',
    },
    body: {
      fontSize: `${baseFontSize}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    subHeading: {
      fontSize: `${baseFontSize * 1.125}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    subHeadingLarge: {
      fontSize: `${baseFontSize * 1.25}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    header: {
      small: {
        fontSize: `${baseFontSize * 1.5}px`,
        lineHeight: `${baseFontSize * 2}px`,
        fontWeight: 'normal',
      },
      medium: {
        fontSize: `${baseFontSize * 1.75}px`,
        lineHeight: `${baseFontSize * 2}px`,
        fontWeight: 'normal',
      },
      large: {
        fontSize: `${baseFontSize * 2}px`,
        lineHeight: `${baseFontSize * 2.5}px`,
        fontWeight: 'normal',
      },
      extraLarge: {
        fontSize: `${baseFontSize * 2.25}px`,
        lineHeight: `${baseFontSize * 2.5}px`,
        fontWeight: 'normal',
      },
    },
    display: {
      small: {
        fontSize: `${baseFontSize * 2.5}px`,
        lineHeight: `${baseFontSize * 3}px`,
        fontWeight: 'normal',
      },
      medium: {
        fontSize: `${baseFontSize * 3}px`,
        lineHeight: `${baseFontSize * 3.25}px`,
        fontWeight: 'normal',
      },
      large: {
        fontSize: `${baseFontSize * 3.5}px`,
        lineHeight: `${baseFontSize * 4}px`,
        fontWeight: 'normal',
      },
      extraLarge: {
        fontSize: `${baseFontSize * 4}px`,
        lineHeight: `${baseFontSize * 4.5}px`,
        fontWeight: 'normal',
      },
    },
  },
};
