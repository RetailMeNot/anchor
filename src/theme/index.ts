export { default as colors } from './colors.theme';
export { default as fonts } from './fonts.theme';
export { default as sizes } from './sizes.theme';
export { default as variables } from './variables.theme';

export { NormalizeCSS, GlobalCSS } from './GlobalStyles';
