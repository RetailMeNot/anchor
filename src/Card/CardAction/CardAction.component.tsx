// REACT
import * as React from 'react';
// VENDOR
import classNames from 'classnames';
import styled from 'styled-components';
// THEME
import { fonts } from '../../theme';

export interface CardActionProps {
    className?: string;
    children?: any;
}

const StyledCard = styled.div`
    box-sizing: border-box;
    position: absolute;
    display: flex;
    justify-content: center;
    align-items: center;
    font-family: ${fonts.fontFamily};
    top: 0;
    right: 0;
    width: 2.5rem;
    height: 2.5rem;
    background-color: white;
`;

export const CardAction: React.FunctionComponent<CardActionProps> = ({
    className,
    children,
    ...props
}: CardActionProps): JSX.Element => (
    <StyledCard
        className={classNames('anchor-card-action', className)}
        {...props}
    >
        {children}
    </StyledCard>
);

export default CardAction;
