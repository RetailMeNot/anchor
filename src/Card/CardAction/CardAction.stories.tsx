// REACT
import * as React from 'react';
// STORYBOOK
import { storiesOf } from '@storybook/react';
// VENDOR
import styled, { ThemeProvider } from '@xstyled/styled-components';
// COMPONENTS
import { CardAction } from './CardAction.component';
// README
import * as README from './README.md';
import { RootTheme } from '../../theme';

const StyledStory = styled('div')`
    padding: 2rem 5rem;
`;

storiesOf('Components/Card/CardAction', module)
    .addParameters({
        readme: {
            sidebar: README,
        },
    })
    .add('Default', () => (
        <ThemeProvider theme={RootTheme}>
            <StyledStory>
                <p>CardAction 1</p>
                <CardAction>Children</CardAction>
            </StyledStory>
        </ThemeProvider>
    ));
