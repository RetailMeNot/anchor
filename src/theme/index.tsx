// Theme Files
import { typography } from './typography.theme';
import { ColorsTheme } from './colors.theme';
import { ThemeProvider as XstyledThemeProvider } from '@xstyled/styled-components';
import styled from '@xstyled/styled-components';
import { space as spaceStyles, SpaceProps } from '@xstyled/system';
import { BUTTON_KEY, BUTTON_THEME } from '../Button';
import { INPUT_KEY, INPUT_THEME } from '../Form/Input/utils';
import React from 'react';
import {Button} from '../Button';
import classNames from 'classnames';

export const RootTheme = {
    typography,
    radii: { none: '0', base: '4px', modal: '8px', circular: '1000px' },
    [BUTTON_KEY]: BUTTON_THEME,
    [INPUT_KEY]: INPUT_THEME,
    colors: ColorsTheme,
    fonts: {
        base:
            'Avenir Next, Segoe UI, Helvetica Neue, Helvetica, Roboto, sans-serif',
        mono: `'SFMono-Regular', Consolas, Liberation Mono, Menlo, Courier, monospace`,
    },
    // These are all minimums
    breakpoints: {
        xs: 0,
        sm: 576,
        md: 768,
        lg: 992,
        xl: 1200,
        xxl: 1600,
    },
    // todo: we should likely remove the color.borders colors and just use this
    // or find a way to reference the colors.border colors here
    // e.g. `thin solid borders.base`
    borders: {
        light: `thin solid ${ColorsTheme.borders.light}`,
        base: `thin solid ${ColorsTheme.borders.base}`,
        dark: `thin solid ${ColorsTheme.borders.dark}`,
    },

    skeleton: {
        colorStart: '#E7E7E7',
        colorEnd: '#D3D3D3',
    },
};

// Exporting out @xstyled's ThemeProvider. Only way I could get it to work w/o issues.
// export class ThemeProvider2 extends XstyledThemeProvider {}

interface IterableObject {
    [s: number]: number | string | boolean | IterableObject;
  }

interface DebugProps {
    property: number | string | boolean | IterableObject;
    propertyName: string;
    rootProperty?: boolean;
}

interface StyledDebugProps {
    indent?: number;
}

const StyledDebug = styled('div')<StyledDebugProps>`
    font-family: mono;
    margin-top: 0.5rem;
    color: white;

    .category .category h2 {
        background-color: #ec8947;
    }

    .category .category .box {
        border-color: #ec8947;
    }

    .category .category .category h2 {
        background-color: #b84343;
        color: white;
    }

    .category .category .category .box {
        border-color: #b84343;
    }

    .category .category .category .category h2 {
        background-color: #91cdd2;
        color: black;
    }

    .category .category .category .category .box {
        border-color: #91cdd2;
    }

    .category .category .category .category .category h2 {
        background-color: #567224;
        color: white;
    }

    .category .category .category .category .category .box {
        border-color: #567224;
    }
`;

const StyledH2 = styled('h2')`
    ${spaceStyles}
    background-color: #ffea92;
    color: black;
    margin-bottom: 0;
    margin-top: 0;
    padding: 0.25rem 0.5rem 0.25rem 1rem;
    font-size: 1rem;
    cursor: pointer;
    border-radius: 0 1rem 1rem 0;
`;

interface BoxProps extends SpaceProps {
    show?: boolean;
}

const Box = styled('div')<BoxProps>`
    ${spaceStyles}
    width: 20rem;
    border-left: solid #ffea92 2px;
    margin: 0.5rem;
    ${({show}) => `display: ${show ? 'block' : 'none'}`}
`;

const Category = styled('div')<SpaceProps>`
    ${spaceStyles}
    width: 20rem;
`;


const ColorBox = styled('div')<any>`
    width: 1rem;
    height: 1rem;
    display: inline-block;
    vertical-align: middle;
    margin-left: 0.5rem;
    background-color: ${({backgroundColor}) => backgroundColor};
    border: solid white 1px;
`;

const Debug = ({property, propertyName}: DebugProps) => {
    const [show, setShow] = React.useState(false);

    return(
        <StyledDebug>
            {(property || property === 0)
                ? (
                    typeof property === 'number' ||
                    typeof property === 'string' ||
                    typeof property === 'boolean' ||
                    typeof property === 'function'
                        ? (
                                <>
                                    <span style={{fontWeight: 'bold'}}>{propertyName}: </span>
                                    <span>
                                        {typeof property === 'function'
                                            ? 'function'
                                            : /^#[a-zA-Z0-9]{4,7}/.test(property.toString())
                                                ? (<>
                                                    <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                                    <ColorBox backgroundColor={property} />
                                                    </>)
                                                : <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                        }
                                    </span>
                                </>
                            )
                        :
                            <Category pl={0} className="category">
                                <StyledH2 onClick={() => setShow(!show)}>{propertyName}</StyledH2>
                                <Box pl={2} className="box" show={show}>
                                    {Object.values(property).map((_property, index) => (
                                        <Debug
                                            key={index}
                                            property={_property}
                                            propertyName={Object.getOwnPropertyNames(property)[index]}
                                        />
                                    ))}
                                </Box>
                            </Category>
                    )
                : (
                    <>
                       Empty.
                    </>
                )
            }
        </StyledDebug>
    );
};

const DebugWrapper = styled('div')`
    background-color: rgba(0,0,0,.75);
    border-left-color: transparent;
    position: fixed;
    left: 0;
    top: 0;
    height: 100%;
    overflow: hidden;
    width: 30rem;
    padding: 1rem 2rem 1rem 0;

    &.closed {
        height:2.4rem;
        width: auto;
        padding-right: 1rem;

        .debug-scroll {
            display: none;
        }
    }
`;

const DebugScroll = styled('div')`
    height: 100%;
    overflow-y: auto;
    width: 100%;
`;

interface ThemeProviderProps {
    children?: any;
    debug?: boolean;
    theme?: any;
}

export const ThemeProvider = ({ children, debug, theme }: ThemeProviderProps) => {
    const isOpen = localStorage.themeProviderDebug === 'true';
    const [isOpenState, setIsOpenState] = React.useState(isOpen);

    return (
        <XstyledThemeProvider theme={theme}>
            {debug &&
                <DebugWrapper className={classNames(!isOpenState ? 'closed' : null)}>
                    <Button
                        onClick={() => {
                            localStorage.themeProviderDebug = !isOpenState;
                            setIsOpenState(!isOpenState);
                        }}
                        size="sm"
                    >
                        ThemeProvider Debug
                    </Button>

                    <DebugScroll className="debug-scroll">
                        <Debug property={theme} propertyName="Theme" />
                    </DebugScroll>
                </DebugWrapper>
            }
            {children}
        </XstyledThemeProvider>
    );
};
