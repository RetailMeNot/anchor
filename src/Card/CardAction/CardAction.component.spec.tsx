// VENDOR
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { ThemeProvider } from '@xstyled/styled-components';
// ANCHOR
import { RootTheme } from '../../theme';
import { CardAction } from './CardAction.component';

// TEST SETUP
const subject = (
    <ThemeProvider theme={RootTheme}>
        <CardAction />
    </ThemeProvider>
);
const wrapper = mount(subject);
const component = shallow(subject);

describe('Component: CardAction', () => {
    it('should be defined', () => {
        expect(subject).toBeDefined();
        expect(wrapper).toBeDefined();
        expect(component).toBeDefined();
        const tree = renderer.create(subject).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
