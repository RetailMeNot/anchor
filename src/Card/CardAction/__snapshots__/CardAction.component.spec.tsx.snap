// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: CardAction should be defined 1`] = `
.c0 {
  box-sizing: border-box;
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-family: Avenir Next,Segoe UI,Helvetica Neue,Helvetica,Roboto,sans-serif;
  top: 0;
  right: 0;
  width: 2.5rem;
  height: 2.5rem;
  background-color: white;
  z-index: 1;
}

<div
  className="c0 anchor-card-action"
/>
`;
